src/Catalyst.Core.Modules.Web3/Web3EthApiExtensions.cs

```diff
@@ -0,0 +1,156 @@
+                    if (!api.DeltaResolver.TryResolve(blockNumber, out cid))
+                    {
+                        delta = default;
+                        return false;
+                    }
+
+                    break;
+                default:
+                    throw new ArgumentOutOfRangeException();
+            }
+
+            delta = api.GetDeltaWithCid(cid);
+            return true;
+        }
+
+        public static PublicEntry ToPublicEntry(this IWeb3EthApi api, TransactionForRpc transactionCall, Keccak root)
+        {
+            return new PublicEntry
+            {
+                Nonce = (ulong) api.StateReader.GetNonce(root, transactionCall.From),
+                SenderAddress = transactionCall.From.Bytes.ToByteString(),
+                ReceiverAddress = transactionCall.To?.Bytes.ToByteString() ?? ByteString.Empty,
+                GasLimit = (ulong) transactionCall.Gas.GetValueOrDefault(),
+                GasPrice = transactionCall.GasPrice.GetValueOrDefault().ToUint256ByteString(),
+                Amount = transactionCall.Value.GetValueOrDefault().ToUint256ByteString(), 
+                Data = transactionCall.Data?.ToByteString() ?? ByteString.Empty,
+                Timestamp = Timestamp.FromDateTime(DateTime.UtcNow) //TODO: this should be set by user and signed
+            };
+        }
+
+        public static CallOutputTracer CallAndRestore(this IWeb3EthApi api, TransactionForRpc transactionCall, DeltaWithCid deltaWithCid)
+        {
+            var parentDelta = deltaWithCid.Delta;
+            Keccak root = parentDelta.StateRootAsKeccak();
+
+            if (transactionCall.Gas == null)
+            {
+                transactionCall.Gas = parentDelta.GasLimit;
+            }
+
+            var publicEntry = api.ToPublicEntry(transactionCall, root);
+
+            var newDelta = deltaWithCid.CreateOneOffDelta(publicEntry);
+
+            CallOutputTracer callOutputTracer = new CallOutputTracer();
+
+            api.StateProvider.StateRoot = root;
+            api.Executor.CallAndRestore(newDelta, callOutputTracer);
+            api.StateProvider.Reset();
+            api.StorageProvider.Reset();
+            return callOutputTracer;
+        }
+    }
+
+    public struct DeltaWithCid
+    {
+        public Delta Delta;
+        public Cid Cid;
+
+        /// <summary>
+        /// Creates a delta for one off execution.
+        /// </summary>
+        public Delta CreateOneOffDelta(PublicEntry publicEntry)
+        {
+            Delta newDelta = Delta.Clone();
+
+            newDelta.PreviousDeltaDfsHash = Cid.ToArray().ToByteString();
+            newDelta.CoinbaseEntries.Clear();
+            newDelta.ConfidentialEntries.Clear();
+            newDelta.PublicEntries.Clear();
+            newDelta.PublicEntries.Add(publicEntry);
+            return newDelta;
+        }
+
+        public void Deconstruct(out Delta delta, out Cid cid)
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Consider refactoring this method in order to remove the need for this 'out' modifier.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4585091)